wg_engine: alpha premultiplied bug fix

SergeyLebedkin opened the pull request:
Fixes an alpha premultiplication bug in the WebGPU engine by introducing
a texture preprocessing stage using compute shaders.

The preprocessing handles alpha premultiplication for non-premultiplied textures and
channel shuffling for ARGB format textures to convert them to the universal RGBA format.

Added compute shaders for texture preprocessing (premultiplication and channel shuffling)
Modified fragment shader to remove redundant alpha premultiplication
Updated texture handling to track premultiplication and shuffle state

<img width="1000" height="578" alt="image" src="https://github.com/user-attachments/assets/c75d139e-17da-40b7-9549-b8786f8dbfa7" />

https://github.com/thorvg/thorvg/issues/4041

SergeyLebedkin committed:
wg_engine: alpha premultiplied bug fix

Fixes an alpha premultiplication bug in the WebGPU engine by introducing
a texture preprocessing stage using compute shaders.

The preprocessing handles alpha premultiplication for non-premultiplied textures and
channel shuffling for ARGB format textures to convert them to the universal RGBA format.

Added compute shaders for texture preprocessing (premultiplication and channel shuffling)
Modified fragment shader to remove redundant alpha premultiplication
Updated texture handling to track premultiplication and shuffle state

https://github.com/thorvg/thorvg/issues/4041

wenjieshen reviewed (commented):
Thank you for the contribution.
I tested this on Windows, and none of the examples showed any regressions. The Lottie example also runs at the same FPS as the previous version.

<img width="320.5" height="330" alt="Screenshot 2026-01-07 235840" src="https://github.com/user-attachments/assets/ac265ab6-321e-4d88-858b-2f66fc14ae00" />
<img width="320.5" height="330" alt="Screenshot 2026-01-07 235208" src="https://github.com/user-attachments/assets/1dedd4e5-37d7-4dd4-89ce-9a69407f3a28" />

wenjieshen committed:
wg_engine: move texture prep to update

Move alpha premultiplication and channel-shuffle preprocessing out
of the compositor flush path and into the renderer update phase.

Motivation: the previous commit preserved the alpha bug fix, but it
was originally written before shared bitmap textures lived behind
WgTextureMgr. After rebasing onto the newer texture-manager path, the
fix worked by passing WgTextureEntry through picture render data and
running compute work from compositor flush.

This follow-up keeps the original GPU-side conversion while adapting
it to the current texture lifecycle. Texture entries are queued after
upload or refresh during prepare() and batched once from postUpdate(),
before any render task samples them.

That keeps straight-alpha bitmap data converted before filtering,
without making texture preprocessing part of render staging or forcing
WgRenderDataPicture to retain texture-manager entries.

wenjieshen commented:
<img width="1136" height="784" alt="Screenshot 2026-07-03 at 6 58 01 PM" src="https://github.com/user-attachments/assets/469f0876-eda6-4f84-9c2e-89d33d6ede06" />
The current version still resolves the issue.

wenjieshen commented on src/renderer/gpu_engine/wg/tvgWgTexturePreprocessor.cpp:
```diff
@@ -0,0 +1,191 @@
+    if (requests.empty()) return true;
+    auto initialized = initialize(context);
+    assert(initialized);
+    (void)initialized;
```
Updated.

wenjieshen commented on src/renderer/gpu_engine/wg/tvgWgTexturePreprocessor.cpp:
```diff
@@ -0,0 +1,191 @@
+bool WgTexturePreprocessor::initialize(WgContext& context)
+{
+    if (computeAlphaPremultBgr && computeAlphaPremult && computeBgrSwizzle) return true;
+    if (!context.device) return false;
```
Updated.

tinyjin reviewed (approved):
LGTM (No regression in Web)

hermet commented:
@wenjieshen could you please come up with the thorvg.benchmark result? Thanks.

wenjieshen commented:
> @wenjieshen could you please come up with the thorvg.benchmark result? Thanks.

PNG: 1.0346× default, 1.0329× rotation
JPEG control: 0.9997× default, 1.0021× rotation

Thank you.

hermet commented:
> > @wenjieshen could you please come up with the thorvg.benchmark result? Thanks.
> 
> PNG: 1.0346× default, 1.0329× rotation JPEG control: 0.9997× default, 1.0021× rotation
> 
> Thank you.

@wenjieshen Please share us the full tests result.

wenjieshen commented:
> > > @wenjieshen could you please come up with the thorvg.benchmark result? Thanks.
> > 
> > 
> > PNG: 1.0346× default, 1.0329× rotation JPEG control: 0.9997× default, 1.0021× rotation
> > Thank you.
> 
> @wenjieshen Please share us the full tests result.

Thank you. They are in the same run. I have stored the record.
Test Case | Rate (After/Before)
-- | --
PNG / Default | 1.0346×
PNG / Rotation | 1.0329×
JPEG / Default | 0.9997×
JPEG / Rotation | 1.0021×
Rectangle | 0.9943×
Circle | 0.9867×
Stroke | 1.0256×
Linear Gradient | 1.0145×
Radial Gradient | 1.0084×
Stroked Rectangle | 0.9999×

hermet commented:
| Test | Rate |
|---|---:|
| Lottie | +0.29% |
| BigDonut | -3.56% |
| Flying | +0.52% |
| ManyStar | +4.30% |
| Samurai | -0.17% |
